### PR TITLE
Adding proposal E

### PR DESCRIPTION
**Pull request by sudo-bmitch**
Signed-off-by: Brandon Mitchell <git@bmitch.net>

This is cherry picks from A, B, and D to create a proposal that works with existing registries while allowing an upgrade path with a value add from that upgrade (more efficient queries and minimal artifacts).

### CODE REVIEW

**Comment by dlorenc** [comment]
Thanks @sudo-bmitch! I like it!

File: docs/proposals/PROPOSAL_E.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+Additional annotations should be considered for filtering various artifact types, e.g. signature public key hash, attestation type, and sbom schema.
+
+These annotations should be considered from the current config schema:
+
+- `org.opencontainers.platform.architecture`: CPU architecture for binaries
+- `org.opencontainers.platform.os`: operating system for binaries
+- `org.opencontainers.platform.variant`: variant of the CPU architecture for binaries
+
+Existing `org.opencontainers.image.*` annotations should be reviewed to consider more generic names (e.g. replacing `image` with `manifest` or `artifact`).
```
Comment (jdolitsky) [inline docs/proposals/PROPOSAL_E.md:49]: format police here - could we nest this somewhere under "JSON Schema" ?

File: docs/proposals/PROPOSAL_E.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+- The registry SHOULD support filtering by annotation using the `filter=<url encoded "name=value">` parameter.
+- Multiple filter parameters may be set on a request, results must match all filters.
+- TBD: More advanced query semantics (e.g. glob, regex, prefix, lists, inverse query).
+- The registry SHOULD support sorting by annotation using the `sort=name` parameter.
+- Artifacts without a specified sorting annotation are sorted last.
+- TBD: Reversing the sort order is needed (prepend a `-` to the name, `sortDesc` parameter, align with how others do this).
+- TBD: an HTTP header or an Index annotation is needed for clients to discover when the returned result supported custom sorting and/or filtering.
+
+### Digest Tags
```
Comment (jdolitsky) [inline docs/proposals/PROPOSAL_E.md:153]: format police - nest under "Registry API"

**Review by jdolitsky — CHANGES_REQUESTED** [review]
lgtm with exception of a few formatting things

File: docs/proposals/PROPOSAL_E.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "config": { ... },
+  "layers": [ ... ],
+  "reference": { // new field
+    "mediaType": "application/vnd.oci.image.manifest.v1+json", // any manifest media type
+    "size": 1234,
+    "digest": "sha256:a1a1a1...",
+    "annotations": [
+      // pull up annotations from referenced manifest here
```
Comment (nishakm) [inline docs/proposals/PROPOSAL_E.md:66]: I missed this during the meeting - this is something the client has to do when moving artifacts from new registries to old registries, correct?

Comment (sudo-bmitch) [reply docs/proposals/PROPOSAL_E.md:66]: This one we decided to remove in the meeting. We shouldn't need it and the only annotations getting pulled up will be in the references API output by the registry.

**Review by jdolitsky — APPROVED** [review]

File: docs/proposals/PROPOSAL_E.md [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+#### Ordering and Pagination
+
+- The registry MUST order the results in a way that allows pagination.
+- Adding the `n` query parameter is used to limit the number of entries per index returned.
+- The registry SHOULD return fewer than `n` results when the generated Index would exceed the recommended maximum manifest size.
+- The `Link` HTTP header is included in the response when additional results are available and is set to the URL for the next page of results.
+
+#### Filtering and Sorting
+
+- The registry SHOULD support filtering by annotation using the `filter=<field><op><value>` parameter.
```
Comment (nishakm) [inline docs/proposals/PROPOSAL_E.md:155]: Now that we have a working group for filtering, should we be defining details here?

Comment (sudo-bmitch) [reply docs/proposals/PROPOSAL_E.md:155]: I think we've only proposed it internally, still a long way from having another OCI WG repo, but I'm happy to track these ideas where ever it makes the most sense. I think we can break up the changes recommended by this proposal to near term and longer term. So if we want to push through image-spec changes and define digest tags formally, those should be possible while we still work out how to define this API and come up with a filtering solution.

**Review by nishakm — APPROVED** [review]

**Review by dmcgowan — APPROVED** [review]

File: docs/proposals/PROPOSAL_E.md [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+If a query results in no references found, an empty manifest list MUST be returned.
+
+#### Ordering and Pagination
+
+- The registry MUST order the results in a way that allows pagination.
+- Adding the `n` query parameter is used to limit the number of entries per index returned.
+- The registry SHOULD return fewer than `n` results when the generated Index would exceed the recommended maximum manifest size.
+- The `Link` HTTP header is included in the response when additional results are available and is set to the URL for the next page of results.
+
+#### Filtering and Sorting
```
Comment (michaelb990) [inline docs/proposals/PROPOSAL_E.md:153]: This is the main part I'm getting tripped up on. I think that allowing filtering and sorting on any attribute ends up complicating the implementation of this unnecessarily. I also think that having this as a SHOULD will make for a varied and unpredictable experience for clients when moving between registries and probably cause bugs due to difficulties in testing. In other proposals we looked at explicitly requiring filtering or ordering on specific attributes only. Could we specify a subset of annotations or fields that registries should be prepared to index on and clients can expect to work?

Comment (sudo-bmitch) [reply docs/proposals/PROPOSAL_E.md:153]: Individual registries could limit themselves to only a subset of annotations, and a client seeing that not all annotations were filtered or sorted would then been required to handle the extra processing themselves. Clients validate the filters/sorting provided in the `org.opencontainers.references.params` annotation, so going between registries should be handled correctly. It won't be two code paths for clients, but an optimization to skip a path when a registry indicates it's already done that work.

Comment (justincormack) [reply docs/proposals/PROPOSAL_E.md:153]: This synthetic index does not fit well with the registry data model, it has no stable URL, is not versioned, no stable hash so no content addressability, and cannot be referred to by another registry object.

Comment (dlorenc) [reply docs/proposals/PROPOSAL_E.md:153]: I think the consensus yesterday (Brandon was out for it) was to drop filtering from this proposal for now.

Comment (sajayantony) [reply docs/proposals/PROPOSAL_E.md:153]: Should consider the schema that proposal A has  - https://github.com/opencontainers/wg-reference-types/blob/main/docs/proposals/PROPOSAL_A.md#api-path @sudo-bmitch

**Pull request merged**